### PR TITLE
Refactor Online DPO

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR refactors the `OnlineDPOTrainer` to have an API that is closer to that of the offline`DPOTrainer`. It also introduces a `LogCompletionsCallback` that produces a table of completions on WandB

### CODE REVIEW

**Comment by qgallouedec** [comment]
After this one is merged https://github.com/huggingface/trl/pull/1598 we'll probably only need to do 

```python
 trainer = OnlineDPOTrainer(..., judge=OpenAIJudge())
```

**Comment by qgallouedec** [comment]
to eval with gpt:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig
from datasets import load_dataset
from trl import OpenAIPairwiseJudge

judge = OpenAIPairwiseJudge("gpt-4o-mini", max_requests=None)

batch_size = 20
num_samples = 200

dataset = load_dataset("trl-internal-testing/tldr-preference-sft-trl-style", split="validation")
dataset = dataset.select(range(num_samples))

prompts = dataset["prompt"]
ref_summmaries = dataset["summary"]
generation_config = GenerationConfig(max_new_tokens=53, min_new_tokens=53)

model_ids = [
    "online_dpo/checkpoint-69",
    "online_dpo/checkpoint-138",
    "online_dpo/checkpoint-207",
    "online_dpo/checkpoint-276",
    "online_dpo/checkpoint-345",
    "online_dpo/checkpoint-414",
    "online_dpo/checkpoint-483",
    "online_dpo/checkpoint-552",
    "online_dpo/checkpoint-621",
    "online_dpo/checkpoint-684",
]

for model_id in model_ids:
    # Load model
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    model = AutoModelForCausalLM.from_pretrained(model_id).to("cuda")
    tokenizer.pad_token = tokenizer.eos_token

    pred_summaries = []
    for idx in range(0, len(prompts), batch_size):
        batch = prompts[idx : idx + batch_size]
        tokenized_batch = tokenizer(batch, return_tensors="pt", padding=True, truncation=True).to(model.device)
        generations = model.generate(**tokenized_batch, generation_config=generation_config)
        for prompt, generation in zip(tokenized_batch.input_ids, generations):
            generation = generation[len(prompt) :]
            summary = tokenizer.decode(generation, skip_special_tokens=True)
            pred_summaries.append(summary)

    judgements = judge.judge(prompts, list(zip(ref_summmaries, pred_summaries)))
    result = 100 * sum([1 for judgement in judgements if judgement == 1]) / 200
    print(f"{model_id}: {result:.2f}")
```

File: examples/scripts/dpo_online.py
Comment (lewtun) [inline examples/scripts/dpo_online.py:1]: I renamed it with `dpo_` prefix to gather the script with the other dpo ones

File: trl/trainer/callbacks.py [hunk 1]
```diff
@@ -233,3 +235,86 @@ def on_evaluate(self, args: TrainingArguments, state: TrainerState, control: Tra
...
+    ```python
+    prompts = ["The capital of France is", "The opposite of up is"]
+    trainer = DPOTrainer(..., callbacks=[LogCompletionsCallback(prompts)])
+    ```
+
+    Args:
+        prompts (`List[str]`):
+            The prompts to generate completions for.
+        freq (`Optional[int]`, *optional*, defaults to `None`):
+            The frequency at which to log completions. If not provided, defaults to `logging_steps`.
```
Comment (lewtun) [inline trl/trainer/callbacks.py:254]: This was previously set as `save_steps` for logging, but I think `logging_steps` is more intuitive

Comment (qgallouedec) [reply trl/trainer/callbacks.py:254]: Yes, but I think you have to be careful here:
- Unlike datapoints, having more than 10 completion batches is completely indigestible, and the number of datapoints for a run is always well over 10.
- WandB tables have a limit of 10,000 rows. For 8 prompts, this means that the run will fail after ~1100 logging steps. 
- Also, I haven't found a way of adding a row, so I have to upload the complete updated table each time. The way wandb works internally is that it stores all logged tables as artifacts. Hence, each run therefore could become very large.

All this to say that I don't know what the best option is here

Comment (lewtun) [reply trl/trainer/callbacks.py:254]: I see, I guess for long runs the user can tweak `logging_steps` accordingly. Let's leave it like this for now and work on a fix if issues are opened

**Comment by lewtun** [comment]
I've tested the code works as intended for the TL;DR experiments of 1B and 2.8B:

```
accelerate launch --config_file examples/accelerate_configs/multi_gpu.yaml \
    examples/scripts/dpo_online.py \
    --model_name_or_path trl-lib/pythia-1b-deduped-tldr-sft  \
    --reward_model_path trl-lib/pythia-1b-deduped-tldr-rm \
    --dataset_name trl-lib/tldr \
    --learning_rate 5.0e-7 \
    --output_dir pythia-1b-deduped-tldr-online-dpo \
    --beta 0.1 \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 2 \
    --num_train_epochs 3 \
    --max_new_tokens 53 \
    --warmup_ratio 0.1 \
    --missing_eos_penalty 1.0 \
    --logging_steps 20 \
    --save_steps 0.1 \
    --push_to_hub
```

and

```
accelerate launch --config_file examples/accelerate_configs/deepspeed_zero2.yaml \
    examples/scripts/dpo_online.py \
    --model_name_or_path trl-lib/pythia-2.8b-deduped-tldr-sft  \
    --reward_model_path trl-lib/pythia-2.8b-deduped-tldr-rm \
    --dataset_name trl-lib/tldr \
    --learning_rate 5.0e-7 \
    --output_dir pythia-2.8b-deduped-tldr-online-dpo \
    --beta 0.1 \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 2 \
    --num_train_epochs 3 \
    --max_new_tokens 53 \
    --warmup_ratio 0.1 \
    --missing_eos_penalty 1.0 \
    --bf16 \
    --logging_steps 20 \
    --save_steps 0.1 \
    --push_to_hub \
```

What is currently not working is the ZeRO-3, which hangs at the generation step of the logcompletions callback. It should be easy to fix this using a logic similar to `unwrap_model_for_generation()`

File: examples/scripts/dpo_online.py [hunk 2]
```diff
@@ -66,19 +80,20 @@
...
 
-    # Compute that only on the main process for faster data processing.
-    # see: https://github.com/huggingface/trl/pull/1255
     with PartialState().local_main_process_first():
-        dataset = dataset.map(process, num_proc=training_args.dataset_num_proc)
+        dataset = dataset.map(prepare_dataset, num_proc=training_args.dataset_num_proc)
 
     prompts = dataset[args.dataset_test_split]["prompt"][:8]
-    log_completions_callback = LogCompletionsCallback(prompts)
+    log_completions_callback = LogCompletionsCallback(prompts, freq=training_args.logging_steps)
```
Comment (qgallouedec) [inline examples/scripts/dpo_online.py:96]: I'd revert this change.
Having this makes it not compatible with logging_steps < 1.
When you set `freq=None`, the callback will get the "true" logging steps from the trainer state, see transformers.Trainer source code:

```python
        if args.logging_steps is not None:
            if args.logging_steps < 1:
                self.state.logging_steps = math.ceil(max_steps * args.logging_steps)
            else:
                self.state.logging_steps = args.logging_steps
```

Comment (lewtun) [reply examples/scripts/dpo_online.py:96]: Ah nice catch! I'll revert

File: docs/source/online_dpo_trainer.md [hunk 3]
```diff
@@ -82,13 +82,14 @@ The logged metrics are as follows. Here is an example [tracked run at Weights an
...
 * `objective/scores_margin`: The mean score margin (according to the external reward model) between the chosen and rejected completions.
+* `rewards/accuracies`: The accuracies of the online DPO's implicit reward model.
 * `rewards/chosen`: The mean reward (according to online DPO's implicit reward model)of the chosen completions.
 * `rewards/rejected`: The mean reward (according to online DPO's implicit reward model) of the rejected completions.
-* `rewards/accuracies`: The accuracies of the online DPO's implicit reward model.
 * `rewards/margins`: The mean reward margin (according to online DPO's implicit reward model) between the chosen and rejected completions.
 * `logps/chosen`: The mean log probabilities of the chosen completions.
 * `logps/rejected`: The mean log probabilities of the rejected completions.
-* `val/contain_eos_token`: The fraction of completions which contain an EOS token.
+* `val/contain_eos+token`: The fraction of completions which contain and EOS token.
```
Comment (qgallouedec) [inline docs/source/online_dpo_trainer.md:91]:
```suggestion
* `val/contain_eos_token`: The fraction of completions which contain and EOS token.
```

File: docs/source/online_dpo_trainer.md [hunk 4]
```diff
@@ -108,188 +101,104 @@ The logged metrics are as follows. Here is an example [tracked run at Weights an
 * Debugging TIP: `objective/rlhf_reward`: this is the ultimate objective of the RLHF training. If training works as intended, this metric should keep going up.
 * Memory TIP: If you are running out of memory, you can try to reduce the `--per_device_train_batch_size` or increase the `--gradient_accumulation_steps` to reduce the memory footprint.
 * Memory TIP: If you have multiple GPUs, you can also run training with DeepSpeed stage 3 to reduce the memory footprint `accelerate launch --config_file examples/accelerate_configs/deepspeed_zero3.yaml`.
-* Usage TIP: We recommend to use the "EOS trick" via `--non_eos_penalty --stop_token eos`, which replaces the score of completions that do not end with an EOS token with a static scalar penalty `--penalty_reward_value`. This can help the model learn to generate more coherent completions.
+* Usage TIP: We recommend to use the "EOS trick" via `--missing_eos_penalty`, which replaces the score of completions that do not end with an EOS token with a static scalar penalty. This can help the model learn to generate more coherent completions.
```
Comment (qgallouedec) [inline docs/source/online_dpo_trainer.md:104]:
```suggestion
* Usage TIP: We recommend to use the "EOS trick" via `--missing_eos_penalty`, which subtract to the score of completions that do not end with an EOS token a static scalar penalty. This can help the model learn to generate more coherent completions.
```

File: docs/source/online_dpo_trainer.md [hunk 5]
```diff
@@ -108,188 +101,104 @@ The logged metrics are as follows. Here is an example [tracked run at Weights an
...
+Many online implementation details are borrowed from the [`PPOv2Trainer`], which is itself based on the [The N+ Implementation Details of RLHF with PPO: A Case Study on TL;DR Summarization](https://huggingface.co/papers/2403.17031). Here are some additional implementation details:
 
-1. When we turn on the EOS trick (i.e., replacing the score of completions that do not end with an EOS token with a scalar penalty score like `-1`) via `--non_eos_penalty --stop_token eos`, it's possible that the chosen and rejected completions have the same score. In this case, we will naively select the completion with the lower index and the chosen completion.
+1. When we turn on the EOS trick (i.e., replacing the score of completions that do not end with an EOS token with a scalar penalty score like `-1`) via `--missing_eos_penalty`, it's possible that the chosen and rejected completions have the same score. In this case, we will naively select the completion with the lower index and the chosen completion.
```
Comment (qgallouedec) [inline docs/source/online_dpo_trainer.md:116]:
```suggestion
```

Since it's now subtracting a scalar, it's now relevant anymore

File: docs/source/online_dpo_trainer.md [hunk 6]
```diff
@@ -108,188 +101,104 @@ The logged metrics are as follows. Here is an example [tracked run at Weights an
...
+    --gradient_checkpointing \
+    --logging_steps 20 \
+    --save_steps 0.1 \
     --push_to_hub
 ```
 
 Checkpoints and experiment tracking are available at:
 
-- [🤗 Model checkpoint](https://huggingface.co/vwxyzjn/ppo_tldr)
+- [🤗 Model checkpoint](https://huggingface.co/collections/trl-lib/online-dpo-66acd3fa38a331a9cd457b07)
```
Comment (qgallouedec) [inline docs/source/online_dpo_trainer.md:187]:
```suggestion
- [🤗 Model checkpoints](https://huggingface.co/collections/trl-lib/online-dpo-66acd3fa38a331a9cd457b07)
```

File: docs/source/online_dpo_trainer.md [hunk 7]
```diff
@@ -108,188 +101,104 @@ The logged metrics are as follows. Here is an example [tracked run at Weights an
...
 - [🐝 Tracked experiment](https://wandb.ai/huggingface/trl/runs/dd2o3g35)
```
[the remark this answers is not in the document]
Comment (lewtun) [reply docs/source/online_dpo_trainer.md:188]: Yep will do once the tl;dr runs finish!

File: trl/trainer/online_dpo_trainer.py [hunk 8]
```diff
@@ -1,590 +1,444 @@
...
-        model.generation_config.eos_token_id = None
-        # generate tokens without truncation / padding
-        model.generation_config.pad_token_id = None
+        if reward_model is not None and judge is not None:
+            warnings.warn(
+                "Both `reward_model` and `judge` are provided. Please choose provide only one of them. "
+                "Ignoring `judge` and using `reward_model`."
+            )
+        elif reward_model is None and judge is None:
+            raise ValueError("Either `reward_model` or `judge` must be provided.")
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:83]: Note for me: add a NotImplementedError if judge is not None. Let's work on integrating judge in another PR.

File: examples/scripts/dpo_online.py (hunk 2 above)
Comment (lewtun) [reply examples/scripts/dpo_online.py]: Reverted in https://github.com/huggingface/trl/pull/1839/commits/8db4b718ff673408901485977b02b1849bc2619e

File: docs/source/online_dpo_trainer.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply docs/source/online_dpo_trainer.md]: Fixed in https://github.com/huggingface/trl/pull/1839/commits/56afe56d2c553930ea355c3711c0fe6ee4dfcdfc

File: trl/trainer/online_dpo_trainer.py [hunk 9]
```diff
@@ -1,590 +1,483 @@
...
-        df = pd.DataFrame(table)
-
-        if self.accelerator.is_main_process:
-            print_rich_table(df.iloc[0 : 0 + 5])
-            if "wandb" in args.report_to:
-                import wandb
-
-                if wandb.run is not None:
-                    wandb.log({"completions": wandb.Table(dataframe=df)})
+    def empty_cache(self):
```
Comment (edbeeching) [inline trl/trainer/online_dpo_trainer.py:472]: This could be moved to utils?

**Review by edbeeching — APPROVED** [review]
One minor comment, otherwise LGTM. Thanks.

File: trl/trainer/online_dpo_trainer.py (hunk 9 above)
Comment (kashif) [reply trl/trainer/online_dpo_trainer.py:472]: also note this helper: https://github.com/huggingface/trl/commit/1e4fb80cbc8b0ad688269f90eac8282203410620

Comment (lewtun) [reply trl/trainer/online_dpo_trainer.py]: Done in https://github.com/huggingface/trl/pull/1839/commits/57eb67338e4a20d41ab6fb33a801fd973bca10c5
